README.md

```diff
@@ -14,11 +16,127 @@ device drivers, even for protcols that are naturally packet-based such as
 e.g. UDP, as the driver packet-extraction logic should do as many sanity
 checks on the bytestream as possible.
 
-## Testing harness
+## Functionality
+
+The main functionalities that `iodrivers_base` provide are:
+
+- infrastructure to convert byte-oriented streams into streams of validated
+  packages
+- library-level test harness
+- generic oroGen integration (for Rock users)
+- built-in support for multiple type of I/O:
+  - serial
+  - udp server: listens on a UDP port. Sends to the IP of the last received
+    UDP message.
+  - udp client: connects to a given IP and port.
+  - tcp client: connects to a given IP and port.
+  - file-based (e.g. for named pipes or Unix sockets)
+
+Note that `iodrivers_base` can (and maybe should) be used even for
+datagram-oriented mediums such as UDP. These mediums do not need
+the packet-reassembly, but do need packet validation.
+
+## Writing a device driver using `iodrivers_base::Driver`
+
+One starts by subclassing `iodrivers_base::Driver` and providing the size of
+the driver's static internal buffer. This buffer should be a multiple of the maximum
+packet size (twice as big is usually enough):
+
+~~~cpp
+class Driver : iodrivers_base::Driver
+{
+    static const int MAX_PACKET_SIZE = 32;
+    static const int INTERNAL_BUFFER_SIZE = MAX_PACKET_SIZE * 4;
+public:
+    Driver();
+
+};
+~~~
+
+~~~cpp
+Driver::Driver()
+    : iodrivers_base::Driver::Driver(INTERNAL_BUFFER_SIZE);
+{
+}
+~~~
+
+The second required step is to implement `iodrivers_base::Driver::extractPacket`.
+This is the method that tells `iodrivers_base`'s internal logic what is meaningful
+data and what isn't.
+
+See below the pseudo-code implementation of most `extractPacket` methods. Note
+that the amount of validation you do in `extractPacket` may vary, but usually
+"more is better", as it is validation that won't have to be done later.
+The return value is fully documented within the method documentation. Check it
+out for a complete specification.
+
+~~~cpp
+int extractPacket(uint8_t const* buffer, int buffer_size) {
+    if (not enough bytes in buffer to contain a start code) {
+        return 0; // wait for new bytes
+    }
+
+    if (!buffer starts with packet start code) {
```

Comment:
For clarity: `if (buffer does not start with a packet start code) {`

Reply:
done